File: src/main/java/com/library/library/BookAppointmentInterFrm.java
package com.library.library;

import com.leewyatt.rxcontrols.controls.RXTextField;
import dao.BookAppointmentDao;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import util.AlertUtil;
import util.AppointmentTable;
import util.Dbutil;
import util.StringUtil;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;

public class BookAppointmentInterFrm {

    private final BookAppointmentDao bookAppointmentDao = new BookAppointmentDao();
    private final Dbutil dbutil = new Dbutil();
    @FXML
    private TableView<AppointmentTable> appointmentTable;
    @FXML
    private TableColumn<AppointmentTable, Integer> b_id;
    @FXML
    private TableColumn<AppointmentTable, String> b_name;
    @FXML
    private TableColumn<AppointmentTable, String> b_type;
    @FXML
    private TextField bidtxt;
    @FXML
    private TextField bookNametxt;
    @FXML
    private CheckBox extendable;
    @FXML
    private TextField idtxt;
    @FXML
    private TableColumn<AppointmentTable, Integer> no;
    @FXML
    private TitledPane operationPanel;
    @FXML
    private CheckBox returned;
    @FXML
    private Button search;
    @FXML
    private Button submit;
    @FXML
    private TableColumn<AppointmentTable, String> t_borrowDate;
    @FXML
    private TableColumn<AppointmentTable, String> t_endDate;
    @FXML
    private TableColumn<AppointmentTable, Integer> t_extendable;
    @FXML
    private TableColumn<AppointmentTable, Integer> t_returned;
    @FXML
    private TableColumn<AppointmentTable, String> t_returnedTime;
    @FXML
    private TableColumn<AppointmentTable, Integer> u_no;
    @FXML
    private RXTextField uid;
    @FXML
    private TextField uidtxt;
    @FXML
    private TableColumn<AppointmentTable, String> userName;

    @FXML
    void initialize() {
        appointmentTable.getSelectionModel().selectedItemProperty()
                .addListener((obs, oldSelection, newSelection) -> {
                    if (newSelection != null) {
                        idtxt.setText(Integer.toString(newSelection.getID()));
                        uidtxt.setText(Integer.toString(newSelection.getUser_ID()));
                        bidtxt.setText(Integer.toString(newSelection.getBook_ID()));
                        bookNametxt.setText(newSelection.getBookName());
                        extendable.setSelected(newSelection.getExtendable() == 1);
                        if (newSelection.getIsReturned() == 1) {
                            returned.setSelected(true);
                            operationPanel.setDisable(true);
                        } else {
                            returned.setSelected(false);
                            operationPanel.setDisable(false);
                        }
                    }
                });
    }

    @FXML
    void deleteText(ActionEvent event) {
        TextField tf = (TextField) event.getSource();
        tf.clear();
    }

    @FXML
    void search(ActionEvent event) {
        if (StringUtil.isEmpty(uid.getText())) {
            AlertUtil.showError("错误", "错误", "请输入用户编号!");
        } else {
            showTableData();
        }
    }

    @FXML
    void submit(ActionEvent event) throws SQLException {
        int id = Integer.parseInt(idtxt.getText());
        int Extendable = extendable.isSelected() ? 1 : 0;
        int isReturned = returned.isSelected() ? 1 : 0;

        int a = AlertUtil.showConfirm("确认", "二次确认", "确认操作?");
        String returnTime = "";
        if (a == 1) {
            Connection conn = getConnection();
            if (returned.isSelected()) {
                returnTime = LocalDateTime.now().toString();
            }
            int b = bookAppointmentDao.update(conn, id, Extendable, isReturned, returnTime);
            if (b > 0) {
                AlertUtil.showAlert("提示", "提示", "修改成功!");
            } else {
                AlertUtil.showError("错误", "错误", "修改失败!");
            }
            resetValue();
        }
    }

    private ObservableList<AppointmentTable> getTableviewList() {
        int uid = Integer.parseInt(this.uid.getText());
        ObservableList<AppointmentTable> TableviewList = FXCollections.observableArrayList();
        Connection conn = getConnection();
        try {
            ResultSet rs = bookAppointmentDao.list(conn, uid);
            while (rs.next()) {
                AppointmentTable appointmentTable = new AppointmentTable(
                        rs.getInt("ID"),
                        rs.getInt("User_ID"),
                        rs.getString("userName"),
                        rs.getInt("Book_ID"),
                        rs.getString("bookTypeName"),
                        rs.getString("bookName"),
                        rs.getString("Start_Time"),
                        rs.getString("End_Time"),
                        rs.getInt("Extendable"),
                        rs.getInt("isReturned"),
                        rs.getString("Return_Time")
                );
                TableviewList.add(appointmentTable);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                dbutil.close(conn);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return TableviewList;
    }

    private void showTableData() {
        ObservableList<AppointmentTable> list = getTableviewList();
        no.setCellValueFactory(new PropertyValueFactory<>("ID"));
        u_no.setCellValueFactory(new PropertyValueFactory<>("User_ID"));
        userName.setCellValueFactory(new PropertyValueFactory<>("userName"));
        b_id.setCellValueFactory(new PropertyValueFactory<>("Book_ID"));
        b_type.setCellValueFactory(new PropertyValueFactory<>("bookTypeName"));
        b_name.setCellValueFactory(new PropertyValueFactory<>("bookName"));
        t_borrowDate.setCellValueFactory(new PropertyValueFactory<>("Start_Time"));
        t_endDate.setCellValueFactory(new PropertyValueFactory<>("End_Time"));
        t_extendable.setCellValueFactory(new PropertyValueFactory<>("Extendable"));
        t_returned.setCellValueFactory(new PropertyValueFactory<>("isReturned"));
        t_returnedTime.setCellValueFactory(new PropertyValueFactory<>("Return_Time"));
        appointmentTable.setItems(list);
    }

    public Connection getConnection() {
        Connection conn = null;
        try {
            conn = dbutil.getConnection();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return conn;
    }

    private void resetValue() {
        idtxt.setText("");
        uidtxt.setText("");
        bidtxt.setText("");
        bookNametxt.setText("");
        extendable.setSelected(false);
        returned.setSelected(false);
        operationPanel.setDisable(true);
        showTableData();
    }
}

File: src/main/java/com/library/library/BookAppointmentInterFrm.kt
package com.library.library

import com.leewyatt.rxcontrols.controls.RXTextField
import com.leewyatt.rxcontrols.event.RXActionEvent
import dao.BookAppointmentDao
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.control.cell.PropertyValueFactory
import util.*
import java.sql.Connection
import java.sql.ResultSet
import java.time.LocalDateTime

class BookAppointmentInterFrm {
    @FXML
    private lateinit var appointmentTable: TableView<AppointmentTable>
    private val dbutil = Dbutil()
    private val bookAppointmentDao = BookAppointmentDao()
    private lateinit var conn: Connection

    @FXML
    private lateinit var b_id: TableColumn<AppointmentTable, Int>

    @FXML
    private lateinit var b_name: TableColumn<AppointmentTable, String>

    @FXML
    private lateinit var b_type: TableColumn<AppointmentTable, String>

    @FXML
    private lateinit var extendable: CheckBox

    @FXML
    private lateinit var no: TableColumn<AppointmentTable, Int>

    @FXML
    private lateinit var returned: CheckBox

    @FXML
    private lateinit var search: Button

    @FXML
    private lateinit var submit: Button

    @FXML
    private lateinit var operationPanel: TitledPane

    @FXML
    private lateinit var uidtxt: TextField

    @FXML
    private lateinit var bidtxt: TextField

    @FXML
    private lateinit var bookNametxt: TextField

    @FXML
    private lateinit var idtxt: TextField

    @FXML
    private lateinit var t_borrowDate: TableColumn<AppointmentTable, String>

    @FXML
    private lateinit var t_endDate: TableColumn<AppointmentTable, String>

    @FXML
    private lateinit var t_extendable: TableColumn<AppointmentTable, Int>

    @FXML
    private lateinit var t_returned: TableColumn<AppointmentTable, Int>

    @FXML
    private lateinit var u_no: TableColumn<AppointmentTable, Int>

    @FXML
    private lateinit var uid: RXTextField

    @FXML
    private lateinit var userName: TableColumn<AppointmentTable, String>

    @FXML
    private lateinit var t_returnedTime: TableColumn<AppointmentTable, String>

    @FXML
    fun initialize() {
        appointmentTable.selectionModel.selectedItemProperty()
            .addListener { _: ObservableValue<out AppointmentTable?>?, _: AppointmentTable?, t1: AppointmentTable? ->
                if (t1 != null) {
                    idtxt.text = t1.id.toString()
                    uidtxt.text = t1.user_ID.toString()
                    bidtxt.text = t1.book_ID.toString()
                    bookNametxt.text = t1.bookName
                    extendable.isSelected = t1.extendable == 1
                    if (t1.isReturned == 1) {
                        returned.isSelected = true
                        operationPanel.isDisable = true
                    } else {
                        returned.isSelected = false
                        operationPanel.isDisable = false
                    }
                }
            }
    }

    @FXML
    fun deleteText(event: RXActionEvent) {
        val tf = event.source as RXTextField
        tf.clear()
    }

    @FXML
    fun search(event: ActionEvent) {
        if (StringUtil.isEmpty(uid.text)) {
            AlertUtil.showError("错误", "错误", "请输入用户编号!")
            return
        } else {
            showTableData()
        }
    }

    @FXML
    fun submit(event: ActionEvent) {
        val id = idtxt.text.toInt()
        var Extendable = 0
        var isReturned = 0
        if (extendable.isSelected) {
            Extendable = 1
        }
        if (returned.isSelected) {
            isReturned = 1
        }
        val a = AlertUtil.showConfirm("确认", "二次确认", "确认操作?")
        var returnTime = ""
        if (a == 1) {
            conn = getConnection()
            if (returned.isSelected) {
                returnTime = LocalDateTime.now().toString()
            }
            val b = bookAppointmentDao.update(conn, id, Extendable, isReturned, returnTime)
            if (b > 0) {
                AlertUtil.showAlert("提示", "提示", "修改成功!")
            } else {
                AlertUtil.showError("错误", "错误", "修改失败!")
            }
            resetValue()
        }
    }

    private fun getTableviewList(): ObservableList<AppointmentTable> {
        val uid = uid.text.toInt()
        val TableviewList = FXCollections.observableArrayList<AppointmentTable>()
        conn = getConnection()
        try {
            val rs: ResultSet = bookAppointmentDao.list(conn, uid)
            var appointmentTable: AppointmentTable
            while (rs.next()) {
                appointmentTable = AppointmentTable(
                    rs.getInt("ID"),
                    rs.getInt("User_ID"),
                    rs.getString("userName"),
                    rs.getInt("Book_ID"),
                    rs.getString("bookTypeName"),
                    rs.getString("bookName"),
                    rs.getString("Start_Time"),
                    rs.getString("End_Time"),
                    rs.getInt("Extendable"),
                    rs.getInt("isReturned"),
                    rs.getString("Return_Time")
                )
                TableviewList.add(appointmentTable)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                dbutil.close(conn)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return TableviewList
    }

    private fun showTableData() {
        val list = getTableviewList()
        no.cellValueFactory = PropertyValueFactory("ID")
        u_no.cellValueFactory = PropertyValueFactory("User_ID")
        userName.cellValueFactory = PropertyValueFactory("userName")
        b_id.cellValueFactory = PropertyValueFactory("Book_ID")
        b_type.cellValueFactory = PropertyValueFactory("bookTypeName")
        b_name.cellValueFactory = PropertyValueFactory("bookName")
        t_borrowDate.cellValueFactory = PropertyValueFactory("Start_Time")
        t_endDate.cellValueFactory = PropertyValueFactory("End_Time")
        t_extendable.cellValueFactory = PropertyValueFactory("Extendable")
        t_returned.cellValueFactory = PropertyValueFactory("isReturned")
        t_returnedTime.cellValueFactory = PropertyValueFactory("Return_Time")
        appointmentTable.items = list
    }

    fun getConnection(): Connection {
        try {
            conn = dbutil.connection
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return conn
    }

    private fun resetValue() {
        idtxt.text = ""
        uidtxt.text = ""
        bidtxt.text = ""
        bookNametxt.text = ""
        extendable.isSelected = false
        returned.isSelected = false
        operationPanel.isDisable = true
        showTableData()
    }
}